### ISSUE TITLE
Stale Tensor during Dataflow Rewrite

### ISSUE THREAD

**Issue by libing4752** [dmlc/tvm#738]
I tried the following test 
 
    data_q = tvm.placeholder(data_shape, name="data_q", dtype=inp_dtype)
    data_m = tvm.placeholder(data_shape, name="data_m", dtype=inp_dtype)
    data_mul_qm = tvm.compute(res_shape,
                       lambda i:  data_q[i].astype(out_dtype) * data_m[i].astype(out_dtype), 
                       name="data_mul_qm")
    data_res = tvm.compute(res_shape,
                      lambda i:  data_mul_qm[i].astype(out_dtype)  ,
                      name="res")
    s = tvm.create_schedule(data_res.op)
    q = s.cache_read(data_q, davinci.scope_out, [data_mul_qm])
    m = s.cache_read(data_m, davinci.scope_out, [data_mul_qm])  
    tmp_mul_qm = s.cache_read(data_mul_qm, davinci.scope_out, [data_res])

and get an error said `cannot find Tensor (shape=[10240], op.name=data_mul_qm) in the inputs of compute(res, 0x18de628)` at line 

    tmp_mul_qm = s.cache_read(data_mul_qm, davinci.scope_out, [data_res])

I read the source code of `cache_read` function in `schedule_dataflow_rewrite.cc`  , find the reason seems the same like #733 , after two cache_read sentences , `data_mul_qm` has changed , but the old ptr in  `data_res` did not update. One way to solve this is to write the `cache_read` before it changes, like 

    tmp_mul_qm = s.cache_read(data_mul_qm, davinci.scope_out, [data_res])
    q = s.cache_read(data_q, davinci.scope_out, [data_mul_qm])
    m = s.cache_read(data_m, davinci.scope_out, [data_mul_qm])  

Is there some other methods? can I just modified `cache_read` function in `schedule_dataflow_rewrite.cc` to just update the contains of `data_mul_qm` rather than create a new one.

**Comment by tqchen** [comment]
In the first case, we should be able to do in current semantics @derisavi do you think if it is relevant to #733 

```python
tmp_mul_qm = s.cache_read(s[data_mul_qm].op.output(0), davinci.scope_out, [data_res])
```

**Comment by tqchen** [comment]
We can create a sugar for this, when we find that ```stage[tensor->op]->op.same_as(tensor->op)```, we map the tensor to be the new tensor so the old semantics also works

**Comment by libing4752** [comment]
so, you mean every time we do cache_read, after changed the tensor ,we scan all the stages , and update it ?

**Comment by tqchen** [comment]
no, instead, everytime when cache_read is called, we detect if the tensor's op get rewritten to another op and change it before we run the replacement

**Comment by tqchen** [comment]
fixed by #738
